<h1 align=center><font size = 5>Best neighborhood in Toronto to open a Restaurant Supply Store</font></h1>

<a id='item1'></a>


In [228]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

#### Extracting raw data from wiki


In [229]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text

In [230]:
soup = BeautifulSoup(source, 'xml')
table=soup.find('table')

In [231]:
#dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
column_names = ['Postalcode','Borough','Neighborhood']
df = pd.DataFrame(columns = column_names)

In [232]:
# Search all the postcode, borough, neighborhood 
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data

In [233]:
df.head()

Postalcode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

#### Cleaning Data


In [234]:
df=df[df['Borough']!='Not assigned']

In [235]:
df.head()

Postalcode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [236]:
temp_df=df.groupby('Postalcode')['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
temp_df=temp_df.reset_index(drop=False)
temp_df.rename(columns={'Neighborhood':'Neighborhood_joined'},inplace=True)

In [237]:
df_clean = pd.merge(df, temp_df, on='Postalcode')
df_clean.drop(['Neighborhood'],axis=1,inplace=True)
df_clean.drop_duplicates(inplace=True)
df_clean.rename(columns={'Neighborhood_joined':'Neighborhood'},inplace=True)

In [238]:
df_clean.head(10)

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5        M9A         Etobicoke      Islington Avenue, Humber Valley Village
6        M1B       Scarborough                               Malvern, Rouge
7        M3B        North York                                    Don Mills
8        M4B         East York              Parkview Hill, Woodbine Gardens
9        M5B  Downtown Toronto                     Garden District, Ryerson

In [239]:
df_clean.shape

(103, 3)

#### Use the cleaned dataframe from problem 1

In [240]:
df_clean.head()

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [241]:
df_clean.shape

(103, 3)

In [242]:
def get_geocode(postal_code):
    # initialize your variable to None
    lat_lng_coords = None
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    return latitude,longitude

#### CSV to create the dataframe

In [243]:
df_geo = pd.read_csv('http://cocl.us/Geospatial_data')

In [244]:
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [245]:
df_geo.rename(columns={'Postal Code':'Postalcode'},inplace=True)
clean_geo = pd.merge(df_geo, df_clean, on='Postalcode')

In [246]:
geo_data = clean_geo[['Postalcode','Borough','Neighborhood', 'Latitude','Longitude']]

In [247]:
geo_data.head(10)

Postalcode      Borough                                     Neighborhood  \
0        M1B  Scarborough                                   Malvern, Rouge   
1        M1C  Scarborough           Rouge Hill, Port Union, Highland Creek   
2        M1E  Scarborough                Guildwood, Morningside, West Hill   
3        M1G  Scarborough                                           Woburn   
4        M1H  Scarborough                                        Cedarbrae   
5        M1J  Scarborough                              Scarborough Village   
6        M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park   
7        M1L  Scarborough                  Golden Mile, Clairlea, Oakridge   
8        M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West   
9        M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

### Explore and cluster the neighborhoods in Toronto

In [248]:
import os
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
import matplotlib.cm as cm
import matplotlib.colors as colors

In [249]:
CLIENT_ID = 'QEF4SFROUUVOQKNPAFNNRDUW4ACAWSYYG312LM3BKEDDKPIZ' # Foursquare ID
CLIENT_SECRET = 'CCBFW1WDVTUYDLPFX3SR0QBB5R2UKRYOZ3F1JDJ3PLQMPDRM' # Foursquare Secret
VERSION = '20180604'

In [250]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [251]:
venues = getNearbyVenues(names= geo_data['Neighborhood'], latitudes=geo_data['Latitude'], longitudes=geo_data['Longitude'])

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

In [252]:
venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                                              Venue  \
0              -79.194353                                            Wendy’s   
1              -79.160497                              Royal Canadian Legion   
2              -79.160497  SEBS Engineering Inc. (Sustainable Energy and ...   
3              -79.188711                                     RBC Royal Bank   
4              -79.188711                                  G & G Electronics   

   Venue Latitude  Venue Longitude              Venue Category  
0       43.807448       -79.199056        Fast Food Restaurant  
1       43.782533       -79.163085                         Bar  
2       43.782371       -79.156820  Construction & Landscaping  
3       43.766790       -79.191151                        Bank  
4       43.765309       -79.191537           Electronics Store

In [253]:
venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      22   
Berczy Park                                                            55   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills                                                              23   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Downsview                                                              16   
Dufferin, Dovercourt Village                                           13   
East Toronto, Broadview North (Old East York)                           3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           70   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Glencairn                                                               4   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
Hillcrest Village                                                       5   
Humber Summit                                                           2   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      5   
India Bazaar, The Beaches West                                         19   
Kennedy Park, Ionview, East Birchmount Park                             4   
Kensington Market, Chinatown, Grange Park                              74   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       13   
Lawrence Park                                                           3   
Leaside                                                                33   
Little Portugal, Trinity                                               45   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor.

In [254]:
# one hot encoding
onehot = pd.get_dummies(venues[['Venue Category']], prefix="", prefix_sep="")
onehot.drop(['Neighborhood'],axis=1,inplace=True) 
onehot.insert(loc=0, column='Neighborhood', value=venues['Neighborhood'] )
onehot.shape

(2136, 273)

In [262]:
grouped = onehot.groupby('Neighborhood').mean().reset_index()

In [263]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [264]:
num_top = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['Neighborhood'] = grouped['Neighborhood']

for ind in np.arange(grouped.shape[0]):
    venues_sorted.iloc[ind, 1:] = return_most_common_venues(grouped.iloc[ind, :], num_top)

venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt        Clothing Store   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Breakfast Spot          Skating Rink   
1        Sandwich Place           Coffee Shop                   Pub   
2                  Bank              Pharmacy         Deli / Bodega   
3   Japanese Restaurant    Chinese Restaurant                  Bank   
4        Sandwich Place           Coffee Shop      Greek Restaurant   

       5th Most Common Venue        6th Most Common Venue  \
0  Latin American Restaurant  Eastern European Restaurant   
1                   Pharmacy                          Gym   
2                Bridal Shop                Shopping Mall   
3             Discount Store          Distribution Center   
4           Sushi Restaurant                    Juice Bar   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0               Dog Run      Doner Restaurant            Donut Shop   
1      Greek Restaurant        Discount Store      Department Store   
2        Sandwich Place                 Diner            Restaurant   
3               Dog Run      Doner Restaurant            Donut Shop   
4                  Café       Thai Restaurant            Restaurant   

      10th Most Common Venue  
0                  Drugstore  
1               Dessert Shop  
2  Middle Eastern Restaurant  
3                Yoga Studio  
4          Indian Restaurant

In [265]:
# set number of clusters
kclusters = 5

grouped_clustering = grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [266]:
# add clustering labels
venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

merged = geo_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
merged = merged.join(venues_sorted.set_index('Neighborhood'), on='Neighborhood')

merged.head()

Postalcode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.194353             4.0        Fast Food Restaurant   
1 -79.160497             1.0  Construction & Landscaping   
2 -79.188711             1.0              Breakfast Spot   
3 -79.216917             1.0                 Coffee Shop   
4 -79.239476             1.0        Caribbean Restaurant   

  2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0   Dumpling Restaurant        Discount Store          Distribution Center   
1                   Bar           Yoga Studio  Eastern European Restaurant   
2     Electronics Store            Restaurant               Medical Center   
3     Korean Restaurant    Mexican Restaurant          Dumpling Restaurant   
4      Hakka Restaurant                Bakery                  Gas Station   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Dog Run      Doner Restaurant            Donut Shop   
1               Dog Run      Doner Restaurant            Donut Shop   
2   Rental Car Location          Intersection    Mexican Restaurant   
3   Distribution Center               Dog Run      Doner Restaurant   
4    Athletics & Sports                  Bank       Thai Restaurant   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0             Drugstore  Eastern European Restaurant          Women's Store  
1             Drugstore          Dumpling Restaurant      Electronics Store  
2                  Bank                  Yoga Studio                Dog Run  
3            Donut Shop                    Drugstore            Yoga Studio  
4   Fried Chicken Joint             Doner Restaurant    Distribution Center

In [267]:
venues_sorted.head()

Cluster Labels                                     Neighborhood  \
0               1                                        Agincourt   
1               1                           Alderwood, Long Branch   
2               1  Bathurst Manor, Wilson Heights, Downsview North   
3               1                                  Bayview Village   
4               1                Bedford Park, Lawrence Manor East   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Clothing Store                Lounge        Breakfast Spot   
1           Pizza Place        Sandwich Place           Coffee Shop   
2           Coffee Shop                  Bank              Pharmacy   
3                  Café   Japanese Restaurant    Chinese Restaurant   
4    Italian Restaurant        Sandwich Place           Coffee Shop   

  4th Most Common Venue      5th Most Common Venue  \
0          Skating Rink  Latin American Restaurant   
1                   Pub                   Pharmacy   
2         Deli / Bodega                Bridal Shop   
3                  Bank             Discount Store   
4      Greek Restaurant           Sushi Restaurant   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Eastern European Restaurant               Dog Run      Doner Restaurant   
1                          Gym      Greek Restaurant        Discount Store   
2                Shopping Mall        Sandwich Place                 Diner   
3          Distribution Center               Dog Run      Doner Restaurant   
4                    Juice Bar                  Café       Thai Restaurant   

  9th Most Common Venue     10th Most Common Venue  
0            Donut Shop                  Drugstore  
1      Department Store               Dessert Shop  
2            Restaurant  Middle Eastern Restaurant  
3            Donut Shop                Yoga Studio  
4            Restaurant          Indian Restaurant

In [269]:
address = 'Toronto, CA'

geo_locator = Nominatim(user_agent="ny_explorer")
location = geo_locator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


## Discussion:

### 5.1 Explaining the results

As we built our list of neighborhoods with Restaurant venues exclusively we discovered most neighborhoods were similar and the greatest concentration of restaurants was in Central Toronto and downtown Toronto. This might seem obvious but it would also appear that these are some of the most affluent neighborhoods in Toronto so there appears to be correlation. By Locating in the general vicinity of the Exact location my friend could be geographically centered in this cluster and poised to service his restaurant customer base with the greatest efficiency.<br>

When we built our our K-Means dataset we used Silhouette analysis to tell us there was a lot of similarity between neighborhoods and the most common restaurants contained with in. Really there was only 2 types of cluster or neighborhoods in greater Toronto. The vast majority of those were in 1 cluster. So Toronto restaurants might be many but they are very homogeneously located near the center of Toronto.<br>

Of the 103 Toronto Neighborhoods gathered only 55.3% or 57 Neighborhoods are above the median after-tax income. 37.8% or 39 Neighborhoods are below he median after-tax income. 6.7% or 7 neighborhoods did not register as it appears their populations are too low. It appears that the greatest concentration of affluence is near central Toronto. We decided to keep all neighborhoods in the dataset regardless of income of population as the majority were close enough.<br> 


## Conclusion:

I feel confident with the recommendation I have given my friend as it is backed up with demonstrated data analysis. While nothing can ever be 100% certain he will certainly be better informed than he was prior to asking for my help.

Much more inference can be obtained with more work. A potential side business for my friend might be assisting new restaurant owners where they might locate a new restaurant, who their competition is and who their clientele might be.